core/sql/executor/HBaseClient_JNI.cpp

```diff
@@ -2844,819 +2871,980 @@ HBC_RetCode HBaseClient_JNI::getHbaseTableInfo(const char* tblName,
-     if (j_cols == NULL)
-     {
-        getExceptionDetails();
-        logError(CAT_SQL_HBASE, __FILE__, __LINE__);
-        logError(CAT_SQL_HBASE, "HTableClient_JNI::startGet()", getLastError());
-        jenv_->DeleteLocalRef(jba_rowID);
-        jenv_->PopLocalFrame(NULL);
-        return HTC_ERROR_GETROWOPEN_PARAM;
-     }  
-     numColsInScan_ = cols.entries();
+  jbyteArray jba_rowID = jenv_->NewByteArray(rowID.len);
+  if (jba_rowID == NULL) {
+    GetCliGlobals()->setJniErrorStr(getErrorText(HBC_ERROR_CHECKANDUPDATEROW_PARAM));
+    if (htc != NULL)
+        NADELETE(htc, HTableClient_JNI, heap);
+    jenv_->PopLocalFrame(NULL);
+    return HBC_ERROR_CHECKANDUPDATEROW_PARAM;
   }
-  else
-     numColsInScan_ = 0;
-  numReqRows_ = 1;
-  jlong j_tid = transID;  
-  jlong j_ts = timestamp;
+  jenv_->SetByteArrayRegion(jba_rowID, 0, rowID.len, (const jbyte*)rowID.val);
   
-  if (hbs_)
-    hbs_->getTimer().start();
-  tsRecentJMFromJNI = JavaMethods_[JM_GET_OPEN].jm_full_name;
-  jint jresult = jenv_->CallIntMethod(javaObj_, 
-            JavaMethods_[JM_GET_OPEN].methodID, j_tid, jba_rowID, 
-            j_cols, j_ts);
-  if (hbs_)
-    {
-      hbs_->incMaxHbaseIOTime(hbs_->getTimer().stop());
-      hbs_->incHbaseCalls();
-    }
+  jobject jRow = jenv_->NewDirectByteBuffer(row.val, row.len);
+  if (jRow == NULL) {
+    GetCliGlobals()->setJniErrorStr(getErrorText(HBC_ERROR_CHECKANDUPDATEROW_PARAM));
+    if (htc != NULL)
+        NADELETE(htc, HTableClient_JNI, heap);
+    jenv_->PopLocalFrame(NULL);
+    return HBC_ERROR_CHECKANDUPDATEROW_PARAM;
+  }
 
-  jenv_->DeleteLocalRef(jba_rowID);  
-  if (j_cols != NULL)
-     jenv_->DeleteLocalRef(j_cols);
+  jbyteArray jba_columnToCheck = jenv_->NewByteArray(columnToCheck.len);
+  if (jba_columnToCheck == NULL) {
+    GetCliGlobals()->setJniErrorStr(getErrorText(HBC_ERROR_CHECKANDUPDATEROW_PARAM));
+    if (htc != NULL)
+        NADELETE(htc, HTableClient_JNI, heap);
+    jenv_->PopLocalFrame(NULL);
+    return HBC_ERROR_CHECKANDUPDATEROW_PARAM;
+  }
+  jenv_->SetByteArrayRegion(jba_columnToCheck, 0, columnToCheck.len, (const jbyte*)columnToCheck.val);
 
-  if (jenv_->ExceptionCheck())
-  {
+  jbyteArray jba_columnValToCheck = jenv_->NewByteArray(columnValToCheck.len);
+  if (jba_columnValToCheck == NULL) {
+    GetCliGlobals()->setJniErrorStr(getErrorText(HBC_ERROR_CHECKANDUPDATEROW_PARAM));
+    if (htc != NULL)
+        NADELETE(htc, HTableClient_JNI, heap);
+    jenv_->PopLocalFrame(NULL);
+    return HBC_ERROR_CHECKANDUPDATEROW_PARAM;
+  }
+  jenv_->SetByteArrayRegion(jba_columnValToCheck, 0, columnValToCheck.len, (const jbyte*)columnValToCheck.val);
+  jlong j_htc = (long)htc;
+  jboolean j_useTRex = useTRex;
+  jlong j_tid = transID;  
+  jlong j_ts = timestamp;
+  jboolean j_asyncOperation = asyncOperation;
+ 
+  if (hbs)
+    hbs->getTimer().start();
+  tsRecentJMFromJNI = JavaMethods_[JM_HBC_DIRECT_CHECKANDUPDATE_ROW].jm_full_name;
```

Comment:
This is useful when there is a crash in JVM. You will be able to figure out which java method was called from JNI at the time of crash from this thread specific variable.